dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusVulkan.java

```diff
@@ -0,0 +1,412 @@
+        for (LitmusVulkanParser.StorageClassSemanticContext scSemantic : storageClassSemanticList) {
+            storageClassSemantics.add(scSemantic.content);
+        }
+        List<String> avvisSemantics = new ArrayList<>();
+        List<LitmusVulkanParser.AvvisSemanticContext> avvisSemanticList = ctx.avvisSemanticList().avvisSemantic();
+        for (LitmusVulkanParser.AvvisSemanticContext avvisSemantic : avvisSemanticList) {
+            avvisSemantics.add(avvisSemantic.content);
+        }
+        if (!mo.equals(Tag.Vulkan.ACQUIRE) && !mo.equals(Tag.Vulkan.RELEASE) && !mo.equals(Tag.Vulkan.ACQ_REL)
+                && !avvis.equals(Tag.Vulkan.AVAILABLE) && !avvis.equals(Tag.Vulkan.VISIBLE)) {
+            throw new ParsingException("Fences must be acquire, release, acq_rel, available or visible");
+        }
+        Fence fence = EventFactory.newFence(ctx.getText().toLowerCase());
+        tagChecker(fence, false, mo, avvis, scope, storageClassSemantics, avvisSemantics);
+        return programBuilder.addChild(mainThread, fence);
+    }
+
+    @Override
+    public Object visitControlBarrier(LitmusVulkanParser.ControlBarrierContext ctx) {
+        String scope = ctx.scope().content;
+        String mo = ctx.mo().content;
+        String avvis = ctx.avvis().content;
+        List<String> storageClassSemantics = new ArrayList<>();
+        List<LitmusVulkanParser.StorageClassSemanticContext> storageClassSemanticList = ctx.storageClassSemanticList().storageClassSemantic();
+        for (LitmusVulkanParser.StorageClassSemanticContext scSemantic : storageClassSemanticList) {
+            storageClassSemantics.add(scSemantic.content);
+        }
+        List<String> avvisSemantics = new ArrayList<>();
+        List<LitmusVulkanParser.AvvisSemanticContext> avvisSemanticList = ctx.avvisSemanticList().avvisSemantic();
+        for (LitmusVulkanParser.AvvisSemanticContext avvisSemantic : avvisSemanticList) {
+            avvisSemantics.add(avvisSemantic.content);
+        }
+        Expression fenceId = (Expression) ctx.barID().accept(this);
+        String fenceIdString = ctx.getText().replace(fenceId.toString(), "");
+        Fence fence = EventFactory.newFenceWithId(fenceIdString.toLowerCase(), fenceId);
+        fence.addTags(Tag.Vulkan.CBAR);
+        if (!mo.equals(Tag.Vulkan.ACQUIRE) && !mo.equals(Tag.Vulkan.RELEASE) && !mo.equals(Tag.Vulkan.ACQ_REL)) {
+            fence.removeTags(Tag.FENCE);
+        }
+        tagChecker(fence, false, mo, avvis, scope, storageClassSemantics, avvisSemantics);
+        return programBuilder.addChild(mainThread, fence);
+    }
+
+    // TODO: Using fences(without 'F' tag) as special device operations for now, need to be changed
+    @Override
+    public Object visitDeviceOperation(LitmusVulkanParser.DeviceOperationContext ctx) {
+        Fence fence = EventFactory.newFence(ctx.getText().toLowerCase());
```

Comment:
I kinda get the feeling we should add a generic visible event to the core that is more general than `Fence` by not having the `F`-tag by default. We could probably even get rid of `Fence` entirely, since it is doesn't have more semantics than its tags either.

Reply:
This sounds good to me